File: CSharp/One/Transforms/InferTypesPlugins/TypeScriptNullCoalesce.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;

namespace One.Transforms.InferTypesPlugins
{
    public class TypeScriptNullCoalesce : InferTypesPlugin {
        public TypeScriptNullCoalesce(): base("TypeScriptNullCoalesce")
        {
            
        }
        
        public override bool canTransform(Expression expr)
        {
            return expr is BinaryExpression binExpr && binExpr.operator_ == "||";
        }
        
        public override Expression transform(Expression expr)
        {
            if (expr is BinaryExpression binExpr2 && binExpr2.operator_ == "||") {
                var litTypes = this.main.currentFile.literalTypes;
                
                binExpr2.left = this.main.runPluginsOn(binExpr2.left);
                var leftType = binExpr2.left.getType();
                
                if (binExpr2.right is ArrayLiteral arrayLit && arrayLit.items.length() == 0) {
                    if (leftType is ClassType classType && classType.decl == litTypes.array.decl) {
                        arrayLit.setActualType(classType);
                        return new NullCoalesceExpression(binExpr2.left, arrayLit);
                    }
                }
                
                if (binExpr2.right is MapLiteral mapLit && mapLit.items.length() == 0) {
                    if (leftType is ClassType classType2 && classType2.decl == litTypes.map.decl) {
                        mapLit.setActualType(classType2);
                        return new NullCoalesceExpression(binExpr2.left, mapLit);
                    }
                }
                
                binExpr2.right = this.main.runPluginsOn(binExpr2.right);
                var rightType = binExpr2.right.getType();
                
                if (binExpr2.right is NullLiteral)
                    // something-which-can-be-undefined || null
                    return binExpr2.left;
                else if (TypeHelper.isAssignableTo(rightType, leftType) && !TypeHelper.equals(rightType, this.main.currentFile.literalTypes.boolean))
                    return new NullCoalesceExpression(binExpr2.left, binExpr2.right);
            }
            return expr;
        }
    }
}

File: PHP/One/Transforms/InferTypesPlugins/TypeScriptNullCoalesce.php
<?php

namespace One\Transforms\InferTypesPlugins\TypeScriptNullCoalesce;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\BinaryExpression;
use One\Ast\Expressions\NullLiteral;
use One\Ast\Expressions\NullCoalesceExpression;
use One\Ast\Expressions\ArrayLiteral;
use One\Ast\Expressions\MapLiteral;
use One\Ast\AstTypes\ClassType;
use One\Ast\AstTypes\TypeHelper;

class TypeScriptNullCoalesce extends InferTypesPlugin {
    function __construct() {
        parent::__construct("TypeScriptNullCoalesce");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof BinaryExpression && $expr->operator === "||";
    }
    
    function transform($expr) {
        if ($expr instanceof BinaryExpression && $expr->operator === "||") {
            $litTypes = $this->main->currentFile->literalTypes;
            
            $expr->left = $this->main->runPluginsOn($expr->left);
            $leftType = $expr->left->getType();
            
            if ($expr->right instanceof ArrayLiteral && count($expr->right->items) === 0) {
                if ($leftType instanceof ClassType && $leftType->decl === $litTypes->array->decl) {
                    $expr->right->setActualType($leftType);
                    return new NullCoalesceExpression($expr->left, $expr->right);
                }
            }
            
            if ($expr->right instanceof MapLiteral && count($expr->right->items) === 0) {
                if ($leftType instanceof ClassType && $leftType->decl === $litTypes->map->decl) {
                    $expr->right->setActualType($leftType);
                    return new NullCoalesceExpression($expr->left, $expr->right);
                }
            }
            
            $expr->right = $this->main->runPluginsOn($expr->right);
            $rightType = $expr->right->getType();
            
            if ($expr->right instanceof NullLiteral)
                // something-which-can-be-undefined || null
                return $expr->left;
            else if (TypeHelper::isAssignableTo($rightType, $leftType) && !TypeHelper::equals($rightType, $this->main->currentFile->literalTypes->boolean))
                return new NullCoalesceExpression($expr->left, $expr->right);
        }
        return $expr;
    }
}
